client/main.lua

```diff
@@ -0,0 +1,233 @@
+  ["duration"] = { "number", nil, true },
+  ["sound"] = { "boolean" },
+  ["position"] = { "table" },
+  ["image"] = { "string" },
+  ["icon"] = { "string" }
+}
+local PersistentNotiMap = {}
+local nuiReady
+
+-- Debug Print Notification
+local function DebugPrintInfo(data, persistent)
+  if cfg.debugMode then
+    print('Notification | Style: ' .. tostring(data.style) .. '\n | Title: ' .. tostring(data.title) .. '\n | Message: ' .. tostring(data.message) .. '\n | Image URL: ' .. tostring(data.image) .. '\n | Icon: ' .. tostring(data.icon) ..'\n | Duration: ' .. tostring(data.duration) .. '\n | Sound: ' .. tostring(data.sound) .. '\n | Custom: ' .. tostring(data.custom) .. '\n | Position: ' .. tostring(data.position) .. '\n | Persistent: ' .. tostring(persistent))
+  end
+end
+
+-- General Debug Print
+local function DebugPrint(msg)
+  if cfg.debugMode then
+    print(msg)
+  end
+end
+
+local function printError(msg)
+  local errMsg = ('^1[T-Notify Error] %s'):format(msg)
+  print(errMsg)
+end
+
+local function verifyTypes(notiTable, isPersistent)
+  local usePersistent
+  for k, v in pairs(NOTI_TYPES) do
+    usePersistent = v[3] and isPersistent or false
+    if not usePersistent and notiTable[k] and v[2] == nil then
+      if notiTable[k] and type(notiTable[k]) ~= v[1] then
+        printError(('Invalid type for %s, expected %s, got %s'):format(k, v[1], type(notiTable[k])))
+        return false
+      end
+    elseif notiTable[k] and v[2] == false then
+      if not notiTable[k] or type(notiTable[k]) ~= v[1] then
+        printError(('Type for %s is nil or is not a %s'):format(k, v[1]))
+        return false
+      end
+    end
+  end
+
+  return true
+end
+
+--Triggers a notification in the NUI using supplied params
+local function SendNotification(data)
+  DebugPrintInfo(data)
+
+  local areTypesValid = verifyTypes(data)
+
+  if areTypesValid then
+    SendNUIMessage(data)
+    if type(sound) == 'table' then
+      PlaySoundFrontend(-1, sound.name, sound.reference, 1)
+    elseif sound == true then
+      PlaySoundFrontend(-1, cfg.sound.name, cfg.sound.reference, 1)
+    end
+  end
+end
+
+--Triggers a notification using persistence
+local function SendPersistentNotification(step, id, options)
+  if debugMode then
+    print('PersistLog | ' ..'\nStep | ' .. step .. '\nID | ' .. id)
+  end
+
+  if not step or not id then
+    return printError('Persistent notifications must have a valid step and id')
+  end
+
+  local areTypesValid = true
+
+  if options then
+    DebugPrintInfo(options, step .. ' ID: ' .. id)
+    areTypesValid = verifyTypes(options, true)
+    if type(options.sound) == 'table' then
+      PlaySoundFrontend(-1, options.sound.name, options.sound.reference, 1)
+    elseif options.sound == true then
+      PlaySoundFrontend(-1, cfg.sound.name, cfg.sound.reference, 1)
+    end
+  end
+
+  if step == 'start' then
+    PersistentNotiMap[id] = true
+  elseif step == 'end' then
+    PersistentNotiMap[id] = false
+  end
```

Comment:
Classic old me; this map manipulation is memory leaky.

```suggestion
  elseif step == 'end' then
    PersistentNotiMap[id] = nil
  end
```